From 2eb3b00b5a6db6e8ece51a1c2553e92bebfb91f3 Mon Sep 17 00:00:00 2001
From: Quentin Casasnovas <quentin.casasnovas@vates.tech>
Date: Tue, 18 Aug 2026 09:47:32 +0200
Subject: [PATCH] intel-i40e: compat: handle newer kernels with dd_key_init
 used indirectly.

There seems to be a compat layer with a HAVE_JUMP_LABEL switch that doesn't
seem set by anything (couldn't find it inside src/kcompat_generator.sh).

On a v4.19.325-cip134 kernel, not defining HAVE_JUMP_LABEL leads to the
following build error:

  /home/builder/rpmbuild/BUILD/intel-i40e-2.25.11/src/i40e_main.c: In function 'i40e_pci_error_resume':
  /home/builder/rpmbuild/BUILD/intel-i40e-2.25.11/src/kcompat_gcc.h:114:36: error: expected expression before '.' token
    /* (struct static_key_false) */ { .key = STATIC_KEY_INIT_FALSE }
				      ^
  ./include/linux/dynamic_debug.h:88:39: note: in definition of macro 'dd_key_init'
   #define dd_key_init(key, init) key = (init)
					 ^
  ./include/linux/dynamic_debug.h:99:2: note: in expansion of macro 'DEFINE_DYNAMIC_DEBUG_METADATA_KEY'
    DEFINE_DYNAMIC_DEBUG_METADATA_KEY(name, fmt, .key.dd_key_false, \
    ^
  ./include/linux/dynamic_debug.h:100:9: note: in expansion of macro 'STATIC_KEY_FALSE_INIT'
	  (STATIC_KEY_FALSE_INIT))
	   ^
  ./include/linux/dynamic_debug.h:133:2: note: in expansion of macro 'DEFINE_DYNAMIC_DEBUG_METADATA'
    DEFINE_DYNAMIC_DEBUG_METADATA(descriptor, fmt);  \
    ^
  ./include/linux/device.h:1473:2: note: in expansion of macro 'dynamic_dev_dbg'
    dynamic_dev_dbg(dev, dev_fmt(fmt), ##__VA_ARGS__)
    ^
  /home/builder/rpmbuild/BUILD/intel-i40e-2.25.11/src/i40e_main.c:18597:2: note: in expansion of macro 'dev_dbg'
    dev_dbg(&pdev->dev, "%s\n", __func__);
    ^
  make[1]: *** [/home/builder/rpmbuild/BUILD/intel-i40e-2.25.11/src/i40e_main.o] Error 1

Force using the HAVE_JUMP_LABEL logic to avoid the build error and use the
code already present to handle this.

More-over the `struct_size` macro already exists in the linux sources, so
was leading to the following warning:

  /home/builder/rpmbuild/BUILD/intel-i40e-2.25.11/src/virtchnl.h:1842:0: warning: "struct_size" redefined [enabled by default]

Signed-off-by: Quentin Casasnovas <quentin.casasnovas@vates.tech>
---
 src/kcompat_gcc.h | 2 +-
 src/virtchnl.h    | 4 +++-
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/kcompat_gcc.h b/src/kcompat_gcc.h
index babfe6da8612..9b9e24acb057 100644
--- a/src/kcompat_gcc.h
+++ b/src/kcompat_gcc.h
@@ -113,7 +113,7 @@
 #define STATIC_KEY_FALSE_INIT \
 	/* (struct static_key_false) */ { .key = STATIC_KEY_INIT_FALSE }
 
-#ifdef HAVE_JUMP_LABEL
+#if 1 || defined(HAVE_JUMP_LABEL)
 /* dd_key_init() is used (indirectly) with arg like "(STATIC_KEY_INIT_FALSE)"
  * from DEFINE_DYNAMIC_DEBUG_METADATA(), which, depending on config has many
  * different definitions (including helper macros).
diff --git a/src/virtchnl.h b/src/virtchnl.h
index cf0a296da3b4..9be2089d6347 100644
--- a/src/virtchnl.h
+++ b/src/virtchnl.h
@@ -1839,8 +1839,10 @@ VIRTCHNL_CHECK_STRUCT_LEN(12, virtchnl_fdir_del);
 
 /* Size of memory required for struct @p followed by an array count of @count
  * of member @member. */
-#define struct_size(p, member, count) 		\
+#ifndef struct_size
+#  define struct_size(p, member, count) 		\
 	(sizeof(*(p)) + count * sizeof(*((p)->member)))
+#endif
 
 #define __vss_full(p, member, count, old)				      \
 	(struct_size(p, member, count) + (old - struct_size(p, member, 0)))
-- 
2.43.0

